src/stactools/core/utils/raster_footprint.py

```diff
@@ -19,76 +22,660 @@
+        """
+        mask = self.data_mask()
+        polygon = self.data_extent(mask)
+        if polygon is None:
+            return None
+        polygon = self.densify_polygon(polygon)
+        polygon = self.reproject_polygon(polygon)
+        polygon = self.simplify_polygon(polygon)
+        return mapping(polygon)  # type: ignore
+
+    def data_mask(self) -> npt.NDArray[np.uint8]:
+        """Produces a mask of valid data in the source image. Nodata pixels
+        values are set to 0, data pixels are set to 1.
+
+        Returns:
+            npt.NDArray[np.uint8]: An 2D array containing 0s and 1s for
+            nodata/data pixels.
+
+        """
+        shape = self.data_array.shape
+        if self.no_data is not None:
+            mask: npt.NDArray[np.uint8] = np.full(shape, 0, dtype=np.uint8)
+            if np.isnan(self.no_data):
+                mask[~np.isnan(self.data_array)] = 1
+            else:
+                mask[np.where(self.data_array != self.no_data)] = 1
+            mask = np.sum(mask, axis=0, dtype=np.uint8)
+            mask[mask > 0] = 1
+        else:
+            mask = np.full(shape, 1, dtype=np.uint8)
+        return mask
+
+    def data_extent(self, mask: npt.NDArray[np.uint8]) -> Optional[Polygon]:
+        """Produces the data footprint in the native CRS.
+
+        Args:
+            mask (npt.NDArray[np.uint8]): A 2D array containing 0s and 1s for
+                nodata/data pixels.
+
+        Returns:
+            Optional[Polygon]: A native CRS polygon of the convex hull of data
+            pixels.
+        """
+        data_polygons = [
+            shape(polygon_dict)
+            for polygon_dict, region_value in rasterio.features.shapes(
+                mask, transform=self.transform
+            )
+            if region_value == 1
+        ]
+
+        if not data_polygons:
+            return None
+        elif len(data_polygons) == 1:
+            polygon = data_polygons[0]
+        else:
+            polygon = MultiPolygon(data_polygons).convex_hull
+
+        return polygon
+
+    def densify_polygon(self, polygon: Polygon) -> Polygon:
+        """Adds vertices to the footprint polygon in the native CRS using either
+        the ``densification_factor`` or ``densification_distance`` specified in
+        the class constructor.
+
+        Args:
+            polygon (Polygon): Footprint polygon in the native CRS.
+
+        Returns:
+            Polygon: Densified footprint polygon in the native CRS.
+        """
+        if self.densification_factor is not None:
+            return Polygon(
+                densify_by_factor(polygon.exterior.coords, self.densification_factor)
+            )
+        elif self.densification_distance is not None:
+            return Polygon(
+                densify_by_distance(
+                    polygon.exterior.coords, self.densification_distance
+                )
+            )
+        else:
+            return polygon
```

Comment:
Should we check to see if both are set, and error if that is the case? I think so...

Reply:
This is done in `__init__()`